Source/EasyNetQ/MessageVersioning/VersionedExchangeDeclareStrategy.cs

```diff
@@ -21,39 +22,38 @@ public VersionedExchangeDeclareStrategy(IConventions conventions, IAdvancedBus a
             this.conventions = conventions;
             this.advancedBus = advancedBus;
 
-            declaredExchanges = new AsyncCache<ExchangeKey, IExchange>((k, c) => advancedBus.ExchangeDeclareAsync(k.Name, k.Type, cancellationToken: c));
+            declaredExchanges = new AsyncCache<ExchangeKey, Exchange>((k, c) => advancedBus.ExchangeDeclareAsync(k.Name, k.Type, cancellationToken: c));
         }
 
         /// <inheritdoc />
-        public Task<IExchange> DeclareExchangeAsync(string exchangeName, string exchangeType, CancellationToken cancellationToken)
+        public Task<Exchange> DeclareExchangeAsync(string exchangeName, string exchangeType, CancellationToken cancellationToken)
         {
             return declaredExchanges.GetOrAddAsync(new ExchangeKey(exchangeName, exchangeType), cancellationToken);
         }
 
         /// <inheritdoc />
-        public Task<IExchange> DeclareExchangeAsync(Type messageType, string exchangeType, CancellationToken cancellationToken)
+        public Task<Exchange> DeclareExchangeAsync(Type messageType, string exchangeType, CancellationToken cancellationToken)
         {
             var messageVersions = new MessageVersionStack(messageType);
             return DeclareVersionedExchangesAsync(messageVersions, exchangeType, cancellationToken);
         }
 
-        private async Task<IExchange> DeclareVersionedExchangesAsync(MessageVersionStack messageVersions, string exchangeType, CancellationToken cancellationToken)
+        private async Task<Exchange> DeclareVersionedExchangesAsync(MessageVersionStack messageVersions, string exchangeType, CancellationToken cancellationToken)
         {
-            IExchange destinationExchange = null;
+            Exchange? destinationExchange = null;
             while (!messageVersions.IsEmpty())
             {
                 var messageType = messageVersions.Pop();
                 var exchangeName = conventions.ExchangeNamingConvention(messageType);
                 var sourceExchange = await DeclareExchangeAsync(exchangeName, exchangeType, cancellationToken).ConfigureAwait(false);
                 if (destinationExchange != null)
-                    await advancedBus.BindAsync(sourceExchange, destinationExchange, "#", cancellationToken).ConfigureAwait(false);
+                    await advancedBus.BindAsync(sourceExchange, destinationExchange.Value, "#", cancellationToken).ConfigureAwait(false);
                 destinationExchange = sourceExchange;
             }
-
-            return destinationExchange;
+            return destinationExchange.Value;
```

Comment:
Will throw if `messageVersions.IsEmpty()` is `true`.

Reply:
You are right, but it's impossible because `messageVersions` contains at least `messageType`. Do you have any ideas how to improve this?

Reply:
Why impossible?

```
var a = new MessageVersionStack(sometype);
a.Pop();
DeclareVersionedExchangesAsync(a, ...)
```

I think you may just `throw InvalidOperationException` in the beginning of `DeclareVersionedExchangesAsync` if `messageVersions.IsEmpty()` as now you can not just return `null`.

Reply:
I mean `DeclareVersionedExchangesAsync` is private, so we have full control on passing MessageVersionStack to it.

Reply:
OK, I did not notice.